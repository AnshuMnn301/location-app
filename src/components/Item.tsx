import React from 'react';
import {View, Text, Button} from 'react-native';
import {ThemeContext} from '../../App';
import {LocationType} from '../utils/apicalls';

type Props = {
  item: LocationType;
  deleteItem: Function;
  section: {
    title: string;
    data: LocationType[];
  };
};

export const transformDate = (date: Date) =>
  `${date.toLocaleDateString()}, ${date.toLocaleTimeString()}`;
export default function Item({item, deleteItem, section}: Props) {
  const styles = React.useContext(ThemeContext);
  const placeName = item?.location;
  const date = new Date(item?.time as number);
  const timeStr = transformDate(date);
  return (
    <View
      style={[
        {
          borderRadius: 10,
          backgroundColor: 'lightgray',
          flexDirection: 'row',
          alignItems: 'center',
        },
        styles.mar,
        styles.pad,
      ]}>
      <View style={{width: '80%'}}>
        <Text style={styles.p1}>{placeName}</Text>
        <Text style={[{marginTop: 5}, styles.p2]}>{timeStr}</Text>
      </View>
      {section.title !== 'Current Location' && (
        <View style={{width: '20%'}}>
          <Button title="remove" onPress={() => deleteItem(item?.id)} />
        </View>
      )}
    </View>
  );
}
