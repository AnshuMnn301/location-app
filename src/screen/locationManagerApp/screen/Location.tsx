import React from 'react';
import {
  View,
  FlatList,
  Button,
  Text,
  SectionList,
  ToastAndroid,
} from 'react-native';
import {
  LocationType,
  requestGeolocationPermission,
  reverseGeoCoding,
} from '../../../utils/apicalls';
import GeoLocation from 'react-native-geolocation-service';
import Item from '../../../components/Item';
import {ThemeContext} from '../../../../App';
import {LocationContext} from '../LocationManagerApp';

export default function Location() {
  const styles = React.useContext(ThemeContext);
  const [, setLocationView] = React.useContext(LocationContext);
  const [places, setPlaces] = React.useState<{
    data: {title: string; data: LocationType[]}[];
    length: number;
  }>({
    data: [
      {
        title: 'Current Location',
        data: [],
      },
      {
        title: 'Previous Location',
        data: [],
      },
    ],
    length: 0,
  });
  React.useEffect(() => {
    let watchId: null | number = null;
    async function getLocationFuncStart() {
      const permissionGanted = await requestGeolocationPermission();
      if (
        permissionGanted?.['android.permission.ACCESS_COARSE_LOCATION'] ===
          'granted' &&
        permissionGanted['android.permission.ACCESS_FINE_LOCATION'] ===
          'granted'
      ) {
        watchId = GeoLocation.watchPosition(
          async position => {
            const {latitude, longitude} = position.coords;
            const place = await reverseGeoCoding(latitude, longitude);
            if (places.data.length <= 30)
              setPlaces(prev => {
                let refData = [...prev.data];
                refData[1].data = [...refData[0].data, ...refData[1].data];
                refData[0].data = [
                  {
                    id: prev.length + 1,
                    location: place.display_name,
                    latlng: {latitude, longitude},
                    time: position.timestamp,
                  },
                ];
                return {
                  data: refData,
                  length: prev.length + 1,
                };
              });
            else GeoLocation.stopObserving();
          },
          error => {
            console.error(error);
          },
          {
            enableHighAccuracy: false,
            distanceFilter: 2000,
            accuracy: {
              android: 'high',
            },
          },
        );
      } else {
        ToastAndroid.show('Error in getting permission', ToastAndroid.SHORT);
      }
    }
    getLocationFuncStart();
    return () => {
      if (watchId !== null) GeoLocation.clearWatch(watchId);
    };
  }, []);

  React.useEffect(() => {
    setLocationView(places.data);
  }, [places.data]);

  const handleDeleteItem = (id = undefined) => {
    if (id === undefined)
      setPlaces(prev => {
        let modData = [...prev.data];
        modData[1].data = [];
        return {
          ...prev,
          data: modData,
        };
      });
    else {
      setPlaces(prev => {
        let modData = [...prev.data];
        modData[1].data = modData[1].data.filter(
          location => location.id !== id,
        );
        return {
          ...prev,
          data: modData,
        };
      });
    }
  };
  return (
    <View style={[{flex: 1}, styles.marV]}>
      <SectionList
        sections={places.data}
        keyExtractor={item => item.id.toString()}
        renderItem={props => <Item {...props} deleteItem={handleDeleteItem} />}
        renderSectionHeader={props => (
          <Text style={[styles.heading, styles.marH]}>
            {props.section.title}
          </Text>
        )}
      />

      <View style={styles.padH}>
        <Button title="Remove All" onPress={() => handleDeleteItem()} />
      </View>
    </View>
  );
}
