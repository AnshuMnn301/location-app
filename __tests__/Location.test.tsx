import 'react-native';
import React from 'react';

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';
import {fireEvent, render, screen} from '@testing-library/react-native'
import Location from '../src/screen/locationManagerApp/screen/Location';
import { Platform } from 'react-native';
import Item from '../src/components/Item';

it('render Item Correctly', () => {
    const tree = renderer.create(<Location/>).toJSON()
    expect(tree).toMatchSnapshot()
})

describe('Testing Functionality', () => {
    beforeEach(() => {
        render(<Location/>)
    })
    it('delete all Locations', () =>{
       fireEvent(screen.getByText('Remove All'), 'onPress',[])
    })
    it('Location list Data key', () =>{
       expect(screen.getByTestId('location-list').props.keyExtractor({id: 1})).toContain(undefined)
    })
    it('Location list render Item', () =>{
        const mockDeleteItem = jest.fn()
        render(<Item deleteItem={mockDeleteItem} section={{"data": [{"id": 3, "latlng": {"latitude": 17.6234567, "longitude": 78.77434}, "location": "South Hillview Drive, Milpitas, Santa Clara County, California, 95035, United States", "time": 1682594913918}, {"id": 2, "latlng": {"latitude": 17.6234567, "longitude": 78.77434}, "location": "South Hillview Drive, Milpitas, Santa Clara County, California, 95035, United States", "time": 1682594867802}, {"id": 1, "latlng": {"latitude": 17.6234567, "longitude": 78.77434}, "location": "Turkapalle, Turkapalle (M) mandal, Yadadri Bhuvanagiri, Telangana, India", "time": 1682594793690}], "title": "Previous Location"}} item={{"id": 3, "latlng": {"latitude": 17.6234567, "longitude": 78.77434}, "location": "Turkapalle, Turkapalle (M) mandal, Yadadri Bhuvanagiri, Telangana, India", "time": 1682594793690}}/>)
        fireEvent(screen.getByText('remove'), 'onPress', [])
        expect(mockDeleteItem).toHaveBeenCalledWith(3)
    })
})