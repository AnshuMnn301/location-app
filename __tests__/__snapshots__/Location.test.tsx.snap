// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render Item Correctly 1`] = `
<View
  style={
    Array [
      Object {
        "flex": 1,
      },
      Object {
        "marginVertical": "2.5%",
      },
    ]
  }
>
  <RCTScrollView
    data={
      Array [
        Object {
          "data": Array [],
          "title": "Current Location",
        },
        Object {
          "data": Array [],
          "title": "Previous Location",
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={
      Array [
        0,
        2,
      ]
    }
    testID="location-list"
  >
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <Text
          style={
            Array [
              Object {
                "fontSize": 24,
                "fontWeight": "bold",
              },
              Object {
                "marginHorizontal": "2.5%",
              },
            ]
          }
        >
          Current Location
        </Text>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      />
      <View
        onLayout={[Function]}
        style={null}
      >
        <Text
          style={
            Array [
              Object {
                "fontSize": 24,
                "fontWeight": "bold",
              },
              Object {
                "marginHorizontal": "2.5%",
              },
            ]
          }
        >
          Previous Location
        </Text>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      />
    </View>
  </RCTScrollView>
  <View
    style={
      Object {
        "paddingHorizontal": "2.5%",
      }
    }
  >
    <View
      accessibilityRole="button"
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Array [
            Object {},
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#007AFF",
                "fontSize": 18,
                "margin": 8,
                "textAlign": "center",
              },
            ]
          }
        >
          Remove All
        </Text>
      </View>
    </View>
  </View>
</View>
`;
